executor/src/witgen/multiplicity_column_generator.rs

```diff
@@ -0,0 +1,208 @@
+            })
+            .collect::<Vec<_>>();
+
+        let all_columns = columns
+            .into_iter()
+            .map(|(name, col)| {
+                let poly_id = self.fixed.try_column_by_name(&name).unwrap();
+                (poly_id, col)
+            })
+            .chain(
+                self.fixed
+                    .fixed_cols
+                    .iter()
+                    // TODO: Avoid clone
+                    // TODO: Find the actual size
+                    .map(|(poly_id, fixed_col)| (poly_id, fixed_col.values_max_size().to_vec())),
+            )
+            .collect::<BTreeMap<_, _>>();
+        let terminal_values = OwnedTerminalValues {
+            trace: all_columns,
+            public_values: publics
+                .into_iter()
+                // Publics might be unavailable if they are later-stage publics.
+                .filter_map(|(k, v)| v.map(|v| (k, v)))
+                .collect(),
+            challenge_values: self.fixed.challenges.clone(),
+        };
+
+        log::trace!(
+            "  Done building trace values, took: {}s",
+            start.elapsed().as_secs_f64()
+        );
+
+        for lookup in phantom_lookups {
+            log::trace!("  Updating multiplicity for phantom lookup: {lookup}");
+            let start = std::time::Instant::now();
+
+            let (rhs_machine_size, rhs_tuples) = self.get_tuples(&terminal_values, lookup.right);
+
+            log::trace!(
+                "    Done collecting RHS tuples, took {}s",
+                start.elapsed().as_secs_f64()
+            );
+            let start = std::time::Instant::now();
+
+            let index = rhs_tuples
+                .iter()
+                .map(|(i, tuple)| {
+                    // There might be multiple identical rows, but it's fine, we can pick any.
+                    (tuple, *i)
+                })
+                .collect::<HashMap<_, _>>();
+
+            log::trace!(
+                "    Done building index, took {}s",
+                start.elapsed().as_secs_f64()
+            );
+            let start = std::time::Instant::now();
+
+            let (_, lhs_tuples) = self.get_tuples(&terminal_values, lookup.left);
+
+            log::trace!(
+                "    Done collecting LHS tuples, took: {}s",
+                start.elapsed().as_secs_f64()
+            );
+            let start = std::time::Instant::now();
+
+            let multiplicity_column_id = lookup.multiplicity_column.unwrap();
+            let multiplicities = multiplicity_columns
+                .entry(multiplicity_column_id)
+                .or_insert_with(|| vec![0; rhs_machine_size]);
+            assert_eq!(multiplicities.len(), rhs_machine_size);
+
+            // Looking up the index is slow, so we do it in parallel.
+            let indices = lhs_tuples
+                .into_par_iter()
+                .map(|(_, tuple)| index[&tuple])
+                .collect::<Vec<_>>();
```

Comment:
Do we have to collect here?

Reply:
I think so, because this is a parallel iterator, which is not a proper iterator.